uid/snowid_test.go

```diff
@@ -0,0 +1,342 @@
+}
+
+func TestBase58(t *testing.T) {
+	assert.Equal(t, len(encodeBase58Map), 58)
+
+	node, err := NewNode(0)
+	if err != nil {
+		t.Fatalf("error creating NewNode, %s", err)
+	}
+
+	for i := 0; i < 10; i++ {
+
+		sf := node.Generate()
+		b58 := sf.String()
+		psf, err := Parse([]byte(b58))
+		if err != nil {
+			t.Fatal(err)
+		}
+		if sf != psf {
+			t.Fatal("Parsed does not match String.")
+		}
+	}
+}
+
+func TestID_JSONUnmarshal(t *testing.T) {
+	obj := struct {
+		ID ID
+	}{}
+
+	node, err := NewNode(122)
+	assert.NilError(t, err)
+	newID := node.Generate()
+
+	source := []byte(`{"id": "` + newID.String() + `"}`)
+
+	err = json.Unmarshal(source, &obj)
+	assert.NilError(t, err)
+
+	assert.Equal(t, newID, obj.ID)
+}
+
+func BenchmarkParse(b *testing.B) {
+	node, _ := NewNode(1)
+	sf := node.Generate()
+	b58 := sf.String()
+
+	b.ReportAllocs()
+
+	b.ResetTimer()
+	for n := 0; n < b.N; n++ {
+		id, err := Parse([]byte(b58))
+		assert.NilError(b, err)
+		assert.Assert(b, id > 0)
+	}
+}
+
+func BenchmarkBase58(b *testing.B) {
+	node, _ := NewNode(1)
+	sf := node.Generate()
+
+	b.ReportAllocs()
+
+	b.ResetTimer()
+	for n := 0; n < b.N; n++ {
+		_ = sf.String()
+	}
+}
+
+func BenchmarkGenerate(b *testing.B) {
+	node, _ := NewNode(1)
+
+	b.ReportAllocs()
+
+	b.ResetTimer()
+	for n := 0; n < b.N; n++ {
+		_ = node.Generate()
+	}
+}
+
+func BenchmarkGenerateMaxSequence(b *testing.B) {
+	nodeBits = 1
+	stepBits = 21
+	node, _ := NewNode(1)
+
+	b.ReportAllocs()
+
+	b.ResetTimer()
+	for n := 0; n < b.N; n++ {
+		_ = node.Generate()
+	}
+}
+
+func TestParse(t *testing.T) {
+	type testCase struct {
+		base58      string
+		expected    ID
+		expectedErr string
+	}
+
+	run := func(t *testing.T, tc testCase) {
+		actual, err := Parse([]byte(tc.base58))
+		if tc.expectedErr != "" {
+			assert.ErrorContains(t, err, tc.expectedErr, "int64=%x", int64(actual))
+			return
+		}
+
+		assert.NilError(t, err)
+		assert.Equal(t, actual, tc.expected, "int64=%x", int64(actual))
+	}
+
+	testCases := []testCase{
+		{
+			base58:   "npL6MjP8Qfc", // 0x7fffffffffffffff
+			expected: ID(0x7fffffffffffffff),
+		},
+		{
+			base58:      "npL6MjP8Qfd", // 0x7fffffffffffffff + 1
+			expectedErr: `invalid base58: value too large`,
+		},
+		{
+			base58:      "JPwcyDCgEuqJPwcyDCgEuq",
+			expectedErr: `invalid base58: too long`,
+		},
+		{
+			base58:      "JPwcyDCgEuq", // 0xffffffffffffffff + 1
+			expectedErr: `invalid base58: value too large`,
+		},
+		{
+			base58:      "self",
+			expectedErr: `invalid base58: byte 2 is out of range`,
+		},
+		{
+			base58:   "4jgmnx8Js8A",
+			expected: 1428076403798048768,
+		},
+		{
+			base58:      "0jgmnx8Js8A",
+			expectedErr: `invalid base58: byte 0 is out of range`,
+		},
+		{
+			base58:      "jgmnxI8Js8A",
+			expectedErr: `invalid base58: byte 5 is out of range`,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.base58, func(t *testing.T) {
+			run(t, tc)
+		})
+	}
+}
+
+func FuzzParse_NoPanic(f *testing.F) {
```

Comment:
The 3 fuzz tests below here are new, and helped find the bugs that were fixed.

In CI it runs the seed cases. As a follow up I'd like to add a nightly job that runs the fuzzer to find new interesting cases.